### PR TITLE
Feat/zero lock reward

### PR DESCRIPTION
**Pull request by aminlatifi**
Added error for zero amount locking
Added condition to not stake/withdraw locking reward, it's the locking reward is zero
Added test for Lock edge cases and errors
Increased fuzzing test rounds to 5000
Replaces `vm.assume` statements with `bound`, for better performance on fuzz testing and fixing issue of lots of global error cases appeared after increasing fuzzing test rounds

### CODE REVIEW

File: contracts/GIVpower.sol [hunk 1]
```diff
@@ -66,6 +68,9 @@ contract GIVpower is GardenUnipoolTokenDistributor, IERC20MetadataUpgradeable {
         if (rounds < 1) {
             revert ZeroLockRound();
         }
+        if (amount == 0) {
```
Comment (0xGabi) [inline contracts/GIVpower.sol:71]: Is there some reason to use `== 0` on this condition and `< 1` on the one above?

File: contracts/GIVpower.sol [hunk 2]
```diff
@@ -87,8 +92,10 @@ contract GIVpower is GardenUnipoolTokenDistributor, IERC20MetadataUpgradeable {
...
 
-        // Add power/farming benefit of locking
-        super.stake(msg.sender, _gainedPowerAmount);
+        if (_gainedPowerAmount > 0) {
+            // Add power/farming benefit of locking
+            super.stake(msg.sender, _gainedPowerAmount);
+        }
 
         emit TokenLocked(msg.sender, amount, rounds, _endRound);
         emit Transfer(address(0), msg.sender, _gainedPowerAmount);
```
Comment (0xGabi) [inline contracts/GIVpower.sol:101]: Is correct to emit the events even when `super.stake` was not called?

File: contracts/GIVpower.sol [hunk 3]
```diff
@@ -121,8 +128,10 @@ contract GIVpower is GardenUnipoolTokenDistributor, IERC20MetadataUpgradeable {
...
 
-            // Reduce power/farming benefit of locking
-            super.withdraw(_account, _releasablePowerAmount);
+            if (_releasablePowerAmount > 0) {
+                // Reduce power/farming benefit of locking
+                super.withdraw(_account, _releasablePowerAmount);
+            }
 
             emit TokenUnlocked(_account, _unlockableTokenAmount, round);
             emit Transfer(_account, address(0), _releasablePowerAmount);
```
Comment (0xGabi) [inline contracts/GIVpower.sol:137]: Same as before. Should we emit events regardless?

File: test/Balance.t.sol [hunk 4]
```diff
@@ -80,7 +77,7 @@ contract BalanceTest is GIVpowerTest {
 
         uint256 lockReward = givPower.calculatePower(amount, rounds) - amount;
 
-        vm.assume(lockReward > 0);
+        //        vm.assume(lockReward > 0);
```
Comment (0xGabi) [inline test/Balance.t.sol:80]: Is this a typo? Or should we remove the line?

File: test/Balance.t.sol [hunk 5]
```diff
@@ -167,28 +164,25 @@ contract BalanceTest is GIVpowerTest {
...
-        vm.assume(amount1 > 0);
-        vm.assume(amount2 > 0);
-        vm.assume(amount1 < MAX_GIV_BALANCE);
-        vm.assume(amount2 < MAX_GIV_BALANCE);
-        vm.assume(amount1 < (MAX_GIV_BALANCE - amount2)); // Same as (amount1 + amount2) < MAX_GIV_BALANCE; to avoid overflow
-
-        vm.assume(amount1 < givPower.calculatePower(amount1, rounds1));
-        vm.assume(amount2 < givPower.calculatePower(amount2, rounds2));
+        // vm.assume(amount1 < givPower.calculatePower(amount1, rounds1));
+        // vm.assume(amount2 < givPower.calculatePower(amount2, rounds2));
```
Comment (0xGabi) [inline test/Balance.t.sol:185]: Were this assumtions not necessary?

File: test/LockRounds.t.sol [hunk 6]
```diff
@@ -48,11 +48,9 @@ contract LockRounds is GIVpowerTest {
 
         vm.assume(rounds > 0);
         vm.assume(rounds <= maxLockRounds);
```
Comment (0xGabi) [inline test/LockRounds.t.sol:50]: Why we are not using `bound` here as well?

**Review by 0xGabi — CHANGES_REQUESTED** [review]
Solid changes, glad we found a few edge cases with a higher fuzz number. Left a few request for change.

File: contracts/GIVpower.sol (hunk 1 above)
Comment (aminlatifi) [reply contracts/GIVpower.sol:71]: No :), which one is preffered?

Comment (0xGabi) [reply contracts/GIVpower.sol:71]: Not a preference, I was wondering.

**Review by 0xGabi — APPROVED** [review]
LGTM!
